mirgecom/flux.py

```diff
@@ -62,15 +150,61 @@ def lfr_flux(cv_tpair, f_tpair, normal, lam):
         object array of :class:`meshmode.dof_array.DOFArray` with outward-pointing
         normals
 
-    lam: :class:`meshmode.dof_array.DOFArray`
+    lam: :class:`~meshmode.dof_array.DOFArray`
+
+        lambda parameter for Lax-Friedrichs/Rusanov flux
+
+    Returns
+    -------
+    numpy.ndarray
+
+        object array of :class:`~meshmode.dof_array.DOFArray` with the
+        Lax-Friedrichs/Rusanov flux.
+    """
+    return flux_lfr(cv_tpair, f_tpair, normal, lam)@normal
+
+
+def flux_lfr(cv_tpair, f_tpair, normal, lam):
```

Comment:
I don't find it super meaningful (and potentially confusing) to offer this functionality of computing the "un-dotted-with-a-normal" flux. Was there a use case? A quick grep suggested it's only used just above here.

Reply:
Leaving it un-dotted lets you pass it in as a parameter to the operator flux routine, e.g.:

```python
flux = inviscid_facial_flux(discr, eos, cv_tpair, flux_func=flux_lfr)
```

If the dot was already applied, I think you would need to define and pass in different functions depending on what is being done on the inside of `inviscid_facial_flux`. Not sure if this applies so much to LFR, but for, say, `flux_central`, you'd need separate versions for `div` and `grad`.

None of this is implemented yet, so this is just an attempt to look ahead a little. Some of what's here will probably end up going away when #502 is implemented (like the separate `divergence_flux_lfr` and `divergence_flux_central`; they can be unified).

I think we could use the `*_facial_flux` convention to distinguish between the two kinds of fluxes we have here (this is already being done in `inviscid.py` and `viscous.py`). I guess really `divergence_flux_*` should be `divergence_facial_flux_*`... but maybe we can save that change for #502. 🤷‍♂️

Reply:
From my perspective integrating *any* quantity over the surface of the element in the surface integral term is going to involve dotting with the facial normals. That surface integral is called a "flux" - and it is separate from the "transport fluxes", or any other quantity that we intend to integrate. I think the different choices come down to how you reconstruct the given quantity at the surface (e.g. LFR, central difference, Roe state, etc). The rest, i.e. dotting or scaling the normal just comes down to whether you are integrating a scalar, or a vector.   Is this a misguided view?

Reply:
> From my perspective integrating _any_ quantity over the surface of the element in the surface integral term is going to involve dotting with the facial normals.

I don't know if you meant to specifically imply "dotting" to mean a dot product -- but some kind of product, I would imagine, yeah.

> That surface integral is called a "flux" - and it is separate from the "transport fluxes", or any other quantity that we intend to integrate.

If we consider the "facial fluxes" in the code to be the things that appear inside the surface integrals, then we can treat them as generalizations of the transport fluxes (w.r.t. a surface). In the case of a conservation law (divergence) they're the same, and in the case of something else like `s = grad(u)`, it generalizes to something new that isn't a `dot(f, n)`. To me `inviscid_facial_flux`, `divergence_flux`, `gradient_flux`, etc., are all computing the same type of thing, just for different operators.

> I think the different choices come down to how you reconstruct the given quantity at the surface (e.g. LFR, central difference, Roe state, etc). The rest, i.e. dotting or scaling the normal just comes down to whether you are integrating a scalar, or a vector. Is this a misguided view?

I think it comes down to what the form of the integral looks like when you convert the volume integral into a surface integral. For a divergence it becomes `dot(u, n)`, for a gradient it becomes `u*n`. I'm not sure what it would look like for, say, a curl. But I don't think it could be either of these.

Reply:
> > That surface integral is called a "flux" - and it is separate from the "transport fluxes", or any other quantity that we intend to integrate.
> 
> If we consider the "facial fluxes" in the code to be the things that appear inside the surface integrals, then we can treat them as generalizations of the transport fluxes (w.r.t. a surface).

I mean in general the surface integral "of some quantity" over the surface of the element (or any closed volume) is, in general, called a flux integral. If you are integrating a vector quantity, the normal component (i.e. that part of the vector sticking in the nhat direction) is obtained by dotting with the normal.

> In the case of a conservation law (divergence) they're the same, and in the case of something else like `s = grad(u)`, it generalizes to something new that isn't a `dot(f, n)`. 

Right, in this case we integrate a *scalar* quantity over the surface. Any scalar we want to compute the "element flux" integral for will integrate the same.  This is why I made the two routines a "scalar" surface integral vs. a "vector" surface integral.

> 
> > I think the different choices come down to how you reconstruct the given quantity at the surface (e.g. LFR, central difference, Roe state, etc). The rest, i.e. dotting or scaling the normal just comes down to whether you are integrating a scalar, or a vector. Is this a misguided view?
> 
> I think it comes down to what the form of the integral looks like when you convert the volume integral into a surface integral. For a divergence it becomes `dot(u, n)`, for a gradient it becomes `u*n`. I'm not sure what it would look like for, say, a curl. But I don't think it could be either of these.

In the case of "curl" operator, the result is a *vector* at the surface and computing it on both sides (+,-) of the face and reconstructing it due to the discontinuity at the face is one part, then for the surface integral it would be dotted with the normal.

Reply:
> In the case of "curl" operator, the result is a _vector_ at the surface and computing it on both sides (+,-) of the face and reconstructing it due to the discontinuity at the face is one part, then for the surface integral it would be dotted with the normal.

This doesn't sound right to me (unless I'm misunderstanding). If you subsequently converted that back into a volume integral, you'd have a divergence.

Reply:
> > In the case of "curl" operator, the result is a _vector_ at the surface and computing it on both sides (+,-) of the face and reconstructing it due to the discontinuity at the face is one part, then for the surface integral it would be dotted with the normal.
> 
> This doesn't sound right to me (unless I'm misunderstanding). If you subsequently converted that back into a volume integral, you'd have a divergence.

I could be wrong here, but I'm viewing this in the context of like a generalized stokes:

volume_integral(d[f]) = surface_integral([f] dA)

If f is a scalar, then d[f] is a gradient
If f is a one-form vector then d[f] is a divergence
If f is a two-form vector then d[f] is a curl

In the surface integral it is always [f]dA.  So that's why I keep saying that in the end it is how the quantity [f] gets reconstructed that is different across systems and use-cases, but the surface integral is always done the same regardless of the physical or numerical content of [f].

FWIW, I'm not arguing against keeping them as-is or continuing along with the operator-specific naming scheme; only offering my perspective on why I named them how they previously were.

Reply:
> > In the case of "curl" operator, the result is a _vector_ at the surface and computing it on both sides (+,-) of the face and reconstructing it due to the discontinuity at the face is one part, then for the surface integral it would be dotted with the normal.
> 
> This doesn't sound right to me (unless I'm misunderstanding). If you subsequently converted that back into a volume integral, you'd have a divergence.

After trying to remind myself how the stokes generalization works, I realized that you are of course right and the way in which the surface integrand is formed is operator-specific.

Reply:
I'm guessing this'll get left as-is and done later via #502?